### ISSUE TITLE
BUG: Freeze after using WAV recorder while playing IQ file

### ISSUE THREAD

**Issue by vladisslav2011** [gqrx-sdr/gqrx#1030]
Steps to reproduce:

1. Start gqrx
2. Open IQ Tool
3. Select some file
4. Click Play
5. Click Rec in audio dock
6. Click Rec in audio dock once more to stop recording
7. Click Play to stop playback
8. Try to close IQ Tool...

Expected behavior:
gqrx switches to hardware device, IQ Tool gets closed, etc.

Observed behavior:
Freeze

Current git master and appimage are affected.
Fixed in https://github.com/gqrx-sdr/gqrx/pull/1013 (just workaround GNU Radio bug by resetting reciever->rx before calling set_input_rate)
Primary cause: incorrect GNU Radio hier_block2_detail implementation...

**Comment by argilo** [comment]
> Primary cause: incorrect GNU Radio hier_block2_detail implementation...

Could you explain this further? Is there a bug in GNU Radio?

**Comment by vladisslav2011** [comment]
Yes. There is a bug in the GNU Radio.
When a hier_block2 output is connected to more than one block, disconnecting at least one of connected blocks, but leaving at least one connected block breaks reference to parent detail in hier_block2_detail, making hier_block2::lock() deadlock in loop.
Disconnecting a block from the hier_block2 output clears hier_block2_detail::d_parent_detail
https://github.com/gnuradio/gnuradio/blob/master/gnuradio-runtime/lib/hier_block2_detail.cc#L309-L312
hier_block2::lock() call hier_block2_detail::lock(), which checks hier_block2_detail::d_parent_detail and calls hier_block2_detail::d_owner->lock() (as hier_block2_detail::d_parent_detail is cleared), which calls hier_block2_detail::lock() and so on. 
https://github.com/gnuradio/gnuradio/blob/master/gnuradio-runtime/lib/hier_block2_detail.cc#L873-L881

**Comment by argilo** [comment]
OK, in that case you should open an issue (and optionally a pull request) in GNU Radio.

**Comment by vladisslav2011** [comment]
I can't reproduce it outside of gqrx.
A simple program, that creates hier_block2 with copy block inside connected to 1 null_source and 2 null_sinks does not freeze when one null_sink gets disconnected. But gqrx freezes in hier_block2::lock() after disconnecting wavfile_sink and trying to stop IQ playback/switch devices. gdb shows main thread stuck jumping between hier_block2::lock and hier_block2_detail::lock with d_parent_detail set to 0.
I have to figure exact conditions, that trigger the bug before opening an issue in GNU Radio tracker.

**Comment by argilo** [comment]
> does not freeze when one null_sink gets disconnected

Do you call `lock()` after disconnecting the block?

**Comment by argilo** [comment]
The line that triggers the infinite recursion is this one, in `nbrx::set_quad_rate(float)`:

https://github.com/gqrx-sdr/gqrx/blob/d916a202742a026d2d3396a5efd87b6408bca068/src/receivers/nbrx.cpp#L109

```
[...]
#10472 0x00007ffff631255d in gr::hier_block2_detail::lock() (this=0x555555cf03f0) at /home/argilo/prefix_310/src/gnuradio/gnuradio-runtime/lib/hier_block2_detail.cc:880
#10473 0x00007ffff631255d in gr::hier_block2_detail::lock() (this=0x555555cf03f0) at /home/argilo/prefix_310/src/gnuradio/gnuradio-runtime/lib/hier_block2_detail.cc:880
#10474 0x00007ffff631255d in gr::hier_block2_detail::lock() (this=0x555555cf03f0) at /home/argilo/prefix_310/src/gnuradio/gnuradio-runtime/lib/hier_block2_detail.cc:880
#10475 0x00007ffff631255d in gr::hier_block2_detail::lock() (this=0x555555cf03f0) at /home/argilo/prefix_310/src/gnuradio/gnuradio-runtime/lib/hier_block2_detail.cc:880
#10476 0x00005555556c1e88 in nbrx::set_quad_rate(float) (this=0x555555967770, quad_rate=1024000) at /home/argilo/prefix_310/src/gqrx/src/receivers/nbrx.cpp:109
#10477 0x0000555555615f70 in receiver::set_input_rate(double) (this=0x555555d394a0, rate=1024000) at /usr/include/c++/9/bits/shared_ptr_base.h:1020
#10478 0x0000555555619e84 in receiver::set_input_device(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) (this=this@entry=0x555555d394a0, device="rtl=0") at /usr/include/c++/9/bits/shared_ptr_base.h:1020
#10479 0x00005555556087a2 in MainWindow::stopIqPlayback() (this=0x7fffffffd750) at /usr/include/c++/9/bits/atomic_base.h:326
#10480 0x00007ffff6b01300 in QMetaObject::activate(QObject*, int, int, void**) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#10481 0x00005555556af40d in CIqTool::on_playButton_clicked(bool) (this=0x555555b95d10, checked=<optimized out>) at /home/argilo/prefix_310/src/gqrx/src/qtgui/iq_tool.cpp:134
[...]
```

**Comment by vladisslav2011** [comment]
Thanks.
It looks, like lock() should be called on hier_block2, not on top_block to trigger the bug.

**Comment by argilo** [comment]
That's great! A simple reproduction test case will be very helpful to the GNU Radio maintainers.

**Comment by argilo** [comment]
I wonder if this bug is what prompted this change (in 2014!):

https://github.com/gqrx-sdr/gqrx/commit/eca0ea7163ff16abd698737fc8ac1baab21be668

**Comment by argilo** [comment]
I made a test case for this as well:
```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

#
# SPDX-License-Identifier: GPL-3.0
#
# GNU Radio Python Flow Graph
# Title: Crash Test
# Author: Clayton Smith
# GNU Radio version: 3.10.0.0-rc2

from gnuradio import blocks
from gnuradio import gr


class hier_test(gr.hier_block2):
    def __init__(self):
        gr.hier_block2.__init__(
            self, "Hierarchical Block Test",
            gr.io_signature(1, 1, gr.sizeof_gr_complex*1),
            gr.io_signature(1, 1, gr.sizeof_gr_complex*1),
        )

        self.blocks_copy_0 = blocks.copy(gr.sizeof_gr_complex*1)

        self.connect((self.blocks_copy_0, 0), (self, 0))
        self.connect((self, 0), (self.blocks_copy_0, 0))


class lock_bug_test(gr.top_block):
    def __init__(self):
        gr.top_block.__init__(self, "Bug Test")

        self.hier_test_0 = hier_test()
        self.blocks_vector_source_x_0 = blocks.vector_source_c((0, 0, 0), False, 1, [])
        self.blocks_null_sink_0 = blocks.null_sink(gr.sizeof_gr_complex*1)
        self.blocks_null_sink_1 = blocks.null_sink(gr.sizeof_gr_complex*1)

        self.connect((self.blocks_vector_source_x_0, 0), (self.hier_test_0, 0))
        self.connect((self.hier_test_0, 0), (self.blocks_null_sink_0, 0))
        self.connect((self.hier_test_0, 0), (self.blocks_null_sink_1, 0))

    def trigger_bug(self):
        self.disconnect((self.hier_test_0, 0), (self.blocks_null_sink_1, 0))
        self.hier_test_0.lock()


def main(top_block_cls=lock_bug_test, options=None):
    tb = top_block_cls()
    tb.trigger_bug()


if __name__ == '__main__':
    main()
```

**Comment by vladisslav2011** [comment]
I've opened an issue: https://github.com/gnuradio/gnuradio/issues/5436

**Comment by vladisslav2011** [comment]
This commit eca0ea7 is not a workaround for the bug. It's workaround for a different issue with the scheduler not knowing what to do with a flowgraph having it's input disconnected (freeze/crash after disconnecting source block from running flowgraph).

**Comment by argilo** [comment]
What makes you think that? The commit message ("lock() / unlock() seems to cause occasioanly hangs") sounds like https://github.com/gnuradio/gnuradio/issues/5436 to me.

**Comment by vladisslav2011** [comment]
I've spent some time trying to trigger this bug wilt top_block::lock()/top_block::unlock() without success: different combination of nested hier_block2's with different numbers of input/output ports, different order of connection/disconnection...
This bug is triggered only and always by calling hier_block2::lock(), not top_block::lock().
This bug is triggered even when the flowgraph is not running, so stopping the flowgraph does not help.
So, occasional freeze after top_block::lock(), that can be avoided by stopping the flowgraph is completely different issue.

**Comment by argilo** [comment]
>So, occasional freeze after top_block::lock(), that can be avoided by stopping the flowgraph is completely different issue.

The commit doesn't just stop the flowgraph, it also removes the `lock()` and `unlock()` calls.

**Comment by argilo** [comment]
I guess they are on the top block though, so the fact that you can't reproduce the issue on the top block might suggest something different was happening.

**Comment by argilo** [comment]
I made a list of all the hierarchical blocks that use `lock()` and `unlock()`, then checked all the places that those blocks could be disconnected from an external block without a subsequent `connect`. I found two scenarios where this occurs:

1. When audio recording stopped: `receiver::stop_audio_recording()`
2. When the AFSK1200 Decoder window is closed: `receiver::stop_sniffer()`

Both of these scenarios trigger https://github.com/gnuradio/gnuradio/issues/5436, leaving the receiver's `d_parent_detail` in an invalid state. A crash or freeze than then be triggered in at least a few ways:

1. Stop the RDS decoder
2. Change the input device
3. Stop the I/Q player

To work around the bug, we can artificially add extra `connect` calls after the `disconnect` calls in `receiver::stop_audio_recording()` and `receiver::stop_sniffer()` to reset `d_parent_detail`. These artificial calls could then be removed once a fix for https://github.com/gnuradio/gnuradio/issues/5436 is widely available.
